src/System.Net.Utilities/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -2,33 +2,268 @@
+                }
+                else
+                {
+                    pr = SendWithPingUtility(address, buffer, timeout, options);
+                }
+            }
+            catch (Exception e)
+            {
+                pingException = e;
+            }
 
-            PingException pe = new PingException(SR.net_ping, new PlatformNotSupportedException());
+            // At this point, either PR has a real PingReply in it, or pingException has an Exception in it.
             var ea = new PingCompletedEventArgs(
-                new PingReply(address, default(PingOptions), default(IPStatus), default(long), buffer), 
-                pe, 
-                false, 
+                pr,
+                pingException,
+                false,
                 asyncOp.UserSuppliedState);
 
             Finish();
             asyncOp.PostOperationCompleted(callback, ea);
         }
 
+        private unsafe PingReply SendIcmpPingOverRawSocket(IPAddress address, byte[] buffer, int timeout, PingOptions options)
+        {
+            EndPoint endPoint = new IPEndPoint(address, 0);
+
+            bool isIpv4 = address.AddressFamily == AddressFamily.InterNetwork;
+            ProtocolType protocolType = isIpv4 ? ProtocolType.Icmp : ProtocolType.IcmpV6;
+            // Use the current thread's ID as the identifier.
+            ushort identifier = (ushort)Environment.CurrentManagedThreadId;
+            IcmpHeader header = new IcmpHeader()
+            {
+                Type = isIpv4 ? (byte)Icmpv4MessageConstants.EchoRequest : (byte)Icmpv6MessageConstants.EchoRequest,
+                Code = 0,
+                HeaderChecksum = 0,
+                Identifier = identifier,
+                SequenceNumber = 0,
+            };
+            byte[] sendBuffer = CreateSendMessageBuffer(header, buffer);
+
+            using (Socket socket = new Socket(address.AddressFamily, SocketType.Raw, protocolType))
+            {
+                socket.ReceiveTimeout = timeout;
+                socket.SendTimeout = timeout;
+                // Setting Socket.DontFragment and .Ttl is not supported on Unix, so ignore the PingOptions parameter.
+
+                int ipHeaderLength = isIpv4 ? IpHeaderLengthInBytes : 0;
+                Stopwatch sw = Stopwatch.StartNew();
+                socket.SendTo(sendBuffer, endPoint);
+                byte[] receiveBuffer = new byte[ipHeaderLength + IcmpHeaderLengthInBytes + buffer.Length];
+                while (sw.ElapsedMilliseconds < timeout)
+                {
+                    int bytesReceived = socket.ReceiveFrom(receiveBuffer, ref endPoint);
+                    byte type, code;
+                    fixed (byte* bytesPtr = receiveBuffer)
+                    {
+                        int icmpHeaderOffset = ipHeaderLength;
+                        IcmpHeader receivedHeader = *((IcmpHeader*)(bytesPtr + icmpHeaderOffset)); // Skip IP header.
```

Comment:
I suppose we should guard against something bad happening here. I can check if I've at least received the ICMP header (IP header length + ICMP header length), otherwise ignore the packet.